[CBRD-25156] integer division considering ORACLE_COMPAT_NUMBER_BEHAVIOR system parameter

hyunikn opened the pull request:
http://jira.cubrid.org/browse/CBRD-25156

ORACLE_COMPAT_NUMBER_BEHAVIOR 시스템 파라메터가 yes 로 설정되어 있을 때
정수 나눗셈 연산에서 operand 들을 우선 numeric 으로 변환한 후 나눗셈의 결과도 numeric 으로 되도록 구현한 것입니다. 
(no 로 설정되어 있을 때는 정수 나눗셈의 결과는 정수)
예를 들면, 1/2 의 결과는 시스템 파라메터 설정에 따라 0.5 (numeric) 또는 0 (integer) 가 됩니다. 

정수 나눗셈 결과가 시스템 설정에 따라 numeric 또는 정수 타입 (short, int, bigint) 가 될 수 있으므로 
정수 나눗셈 연산 표현식의 타입은 이 두 경우를 아우르는 Object 로 타이핑을 할 수 있습니다. 

이번 PR 에서는 연산의 operand 중 하나가 Object 로 타이핑되어 있을 때 실행시간에 타입 검사를 하여 
타입별로 올바른 연산으로 분기하도록 구현하는 부분도 새롭게 추가하였습니다. 

두 개의 commit 중 2606dd2 이 COMPAT_NUMBER_BEHAVIOR 시스템 파라메터를 적용하는 부분이고 (비교적 단순)
6066fac 가 컴파일 시간에 Object 로 타이핑된 operand 들을 실행시간에 타입 검사를 하여 
타입별 연산으로 분기하는 부분을 구현한 것입니다. 

그리고, 기존에 + 연산자가 문자열에 적용되었을 때 double 로 캐스팅한 후 숫자 연산을 하던 것을 
큐브리드 엔진과 마찬가지로 문자열 concat 으로 동작하도록 수정하는 내용도 함께 포함되어 있습니다.

hyunikn committed:
apply ORACLE_COMPAT_NUMBER_BEHAVIOR to division operation of Shorts, Integers, and Bigints

hyunikn committed:
handling operands of OBJECT type (ANY type) of the following operators
  +, -, *, /, div, mod, unary -, ~

hyunikn committed:
Merge branch 'develop' into CBRD-25156-oracle-int-division

vimkim commented on pl_engine/pl_server/src/main/java/com/cubrid/plcsql/predefined/sp/SpLib.java:
```diff
@@ -729,7 +764,42 @@ public static Object opBitCompli(Object l) {
+        } else if (l instanceof Long) {
+            // bigint
+            return opBitCompli((Long) l);
+        } else if (l instanceof BigDecimal) {
+            // not applicable
+            return opBitCompli(convNumericToBigint((BigDecimal) l));
+        } else if (l instanceof Float) {
+            // not applicable
+            return opBitCompli(convFloatToBigint((Float) l));
+        } else if (l instanceof Double) {
+            // not applicable
+            return opBitCompli(convDoubleToBigint((Double) l));
+        } else if (l instanceof Date) {
+            // not applicable
+        } else if (l instanceof Time) {
+            // not applicable
+        } else if (l instanceof Timestamp) {
+            throw new PROGRAM_ERROR("operand's type is ambiguous: TIMESTAMP or DATETIME");
```
BigDecimal, Float, Double 은 주석에 not applicable이라고 적혀있지만 return 값이 존재하는데,
주석을 지워야 하는게 맞는지 궁금합니다.

hyunikn replied:
주석이 틀렸네요. 수정하겠습니다. 
감사합니다.

hyunikn committed:
consider ORACLE_STYLE_EMPTY_STRING in opAdd(String, String)

hyunikn committed:
. update serveral comments
. set common types of comparison operation to OBJECT if either argument is of type OBJECT

hyunikn committed:
minor: capitalize type names in error messages

hgryoo commented:
- cci의 서브모듈 수정사항이 잘못 포함되어 있는 것 같습니다
- 현재 draft PR 상태라 상태 변경 부탁드립니다

hyunikn committed:
fix cubrid-cci version

hyunikn commented:
> * cci의 서브모듈 수정사항이 잘못 포함되어 있는 것 같습니다
> * 현재 draft PR 상태라 상태 변경 부탁드립니다

아. 깜박... ^^ 수정했습니다.

hyunikn commented on pl_engine/pl_server/src/main/java/com/cubrid/plcsql/compiler/CoercionScheme.java:
```diff
@@ -357,41 +360,41 @@ public abstract List<Type> getCoercions(
 
         compOpCommonType[Type.IDX_BOOLEAN][Type.IDX_NULL] = Type.BOOLEAN;
-        compOpCommonType[Type.IDX_BOOLEAN][Type.IDX_OBJECT] = Type.BOOLEAN;
+        compOpCommonType[Type.IDX_BOOLEAN][Type.IDX_OBJECT] = Type.OBJECT;
```
comparison operator =, !=, <, > 등에서 인자 타입이 (Boolean, Object) 였을 때 그들의 common 타입을 Boolean 으로 했었으나, 이제는 Object 로 한다는 뜻입니다. 
Common 타입이 Object 이면 실행시간 타입검사를 해서 분기하는 코드로 진행하는 효과가 있습니다.

hyunikn replied:
이하, compOpCommonType, arithOpCommonType, intArithOpCommonType 들은 연산자 인자의 common 타입 관련된 큐브리드의 규칙을 구현한 것이고 이것과 관련해서 해당 Jira 이슈 CBRD-25156에 첨부된 엑셀 파일에서 비교연산, 산술연산, 정수산술연산 탭의 내용을 참고하실 수 있습니다.

hyunikn commented on pl_engine/pl_server/src/main/java/com/cubrid/plcsql/compiler/CoercionScheme.java:
```diff
@@ -498,16 +504,26 @@ public abstract List<Type> getCoercions(
         arithOpCommonType[Type.IDX_DOUBLE][Type.IDX_FLOAT] = Type.DOUBLE;
         arithOpCommonType[Type.IDX_DOUBLE][Type.IDX_DOUBLE] = Type.DOUBLE;
 
+        addCommonTypeExt[Type.IDX_STRING][Type.IDX_STRING] = Type.STRING_ANY;
```
addCommonTypeExt 는 산술연산자 중 + 에만 적용되는 규칙을 구현한 것 입니다. arithOpCommonType 을 확장합니다.

hyunikn commented on pl_engine/pl_server/src/main/java/com/cubrid/plcsql/compiler/CoercionScheme.java:
```diff
@@ -498,16 +504,26 @@ public abstract List<Type> getCoercions(
+        addCommonTypeExt[Type.IDX_DATETIME][Type.IDX_OBJECT] = Type.OBJECT;
+        addCommonTypeExt[Type.IDX_TIMESTAMP][Type.IDX_OBJECT] = Type.OBJECT;
+
+        subtractCommonTypeExt[Type.IDX_DATE][Type.IDX_OBJECT] = Type.OBJECT;
```
subtractCommonTypeExt 는 산술연산자 중 - 에만 적용되는 규칙을 구현한 것 입니다. arithOpCommonType 을 확장합니다.

hyunikn commented on pl_engine/pl_server/src/main/java/com/cubrid/plcsql/compiler/CoercionScheme.java:
```diff
@@ -607,7 +625,7 @@ private static Type getCommonTypeInner(Type lType, Type rType, Type[][]... table
 
         int tables = table.length;
-        for (int i = 0; i < tables; i++) {
+        for (int i = tables - 1; i >= 0; i--) { // check extensions first
```
기본 규칙보다 확장 규칙을 먼저 찾아보는 것으로 순서가 바뀌었습니다. 
현재 산술 연산에 대한 기본 규칙과 + 에 대한 확장 규칙, - 에 대한 확장 규칙에 적용됩니다. 
산술 연산을 제외하고는 확장 규칙이 없습니다.

hyunikn commented on pl_engine/pl_server/src/main/java/com/cubrid/plcsql/predefined/sp/SpLib.java:
```diff
@@ -689,7 +689,42 @@ public static Object opNeg(Object l) {
             return null;
         }
-        throw new PROGRAM_ERROR(); // unreachable
+
```
인자 타입이 Object 인 메소드에 대해서는 run-time type 을 검사하여 분기하는 코드로 진행합니다. 
인자 타입이 (Object, Object) 인 이항 연산 해당 메소드들도 마찬가지 입니다.

hyunikn commented on pl_engine/pl_server/src/main/java/com/cubrid/plcsql/predefined/sp/SpLib.java:
```diff
@@ -2234,11 +2320,29 @@ public static Object opMod(Object l, Object r) {
 
     // ====================================
     // +
+    @Operator(coercionScheme = CoercionScheme.ArithOp)
```
큐브리드는 문자열에 대한 덧셈 연산을 concat으로 동작하는 것이 기본 동작입니다. (시스템 파라메터로 다르게 정할 수도 있지만)
기존에 + 는 무조건 산술 연산으로만 동작하도록 구현했었는데, 이번 PR 에서 문자열에 대해서는 concat으로 동작하도록 함께 수정했습니다.

hyunikn commented on pl_engine/pl_server/src/main/java/com/cubrid/plcsql/predefined/sp/SpLib.java:
```diff
@@ -689,7 +689,42 @@ public static Object opNeg(Object l) {
+            // not applicable
+        } else if (l instanceof Time) {
+            // not applicable
+        } else if (l instanceof Timestamp) {
```
큐브리드의 DATETIME과 TIMESTAMP가 모두 Java 의 Timestamp 에 대응되기 때문에 Java 객체의 타입을 가지고는 큐브리드 타입을 알 수가 없어서 일단 이런 경우는 모두 에러를 내는 것으로 구현했습니다. 
별도 이슈를 통해 DATETIME과 TIMESTAMP 에 대응하는 Java 타입을 구별하는 것이 좋을 것 같습니다.

hyunikn committed:
trim strings before they are converted to one of number types SHORT, INT, BIGINT, FLOAT, and DOUBLE
(NOTE: no trimming before converting to NUMERIC type)

hyunikn committed:
Merge branch 'develop' into CBRD-25156-oracle-int-division
